src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -944,6 +924,64 @@ private SortVersion GetSortVersion()
                                                              (byte) (LCID  & 0xFF)));
         }
 
+        private static class SortHandleCache
+        {
+            // in most scenarios there is a limited number of cultures with limited number of sort options
+            // so caching the sort handles and not freeing them is OK, see https://github.com/dotnet/coreclr/pull/25117 for more
+            private static readonly Dictionary<string, IntPtr> s_sortNameToSortHandleCache = new Dictionary<string, IntPtr>();
+
+            internal static IntPtr GetCachedSortHandle(string sortName)
+            {
+                lock (s_sortNameToSortHandleCache)
+                {
+                    if (!s_sortNameToSortHandleCache.TryGetValue(sortName, out IntPtr result))
+                    {
+                        result = GetSortHandle(sortName);
+
+                        try
+                        {
+                            s_sortNameToSortHandleCache.Add(sortName, result);
+                        }
+                        catch
+                        {
+                            Interop.Globalization.CloseSortHandle(result);
+
+                            throw;
+                        }
+                    }
+
+                    return result;
+                }
+            }
+
+            private static IntPtr GetSortHandle(string sortName)
+            {
+                switch(Interop.Globalization.GetSortHandle(GetNullTerminatedUtf8String(sortName), out IntPtr sortHandle))
+                {
+                    case Interop.Globalization.ResultCode.Success:
+                        return sortHandle;
+                    case Interop.Globalization.ResultCode.OutOfMemory:
+                        throw new OutOfMemoryException();
+                    default:
+                        throw new ExternalException(SR.Arg_ExternalException);
+                }
+            }
+
+            private static byte[] GetNullTerminatedUtf8String(string s)
```

Comment:
This method is not needed. You can just depend on the default interop marshaling that will convert the string to UTF8.

Reply:
is the default interop marshaling also include the null terminating character?

Reply:
yes